File: src/main/java/com/example/oauth/util/JsonUtils.java
package com.example.oauth.util;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.experimental.UtilityClass;

import java.util.Optional;

@UtilityClass
public class JsonUtils {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    public <T> Optional<T> readValue(String content, Class<T> valueType) {
        try {
            return Optional.of(MAPPER.readValue(content, valueType));
        } catch (JsonProcessingException e) {
            return Optional.empty();
        }
    }

    public <T> Optional<T> readValue(String content, TypeReference<T> valueTypeRef) {
        try {
            return Optional.of(MAPPER.readValue(content, valueTypeRef));
        } catch (JsonProcessingException e) {
            return Optional.empty();
        }
    }

    public Optional<String> writeValue(Object value) {
        try {
            return Optional.of(MAPPER.writeValueAsString(value));
        } catch (JsonProcessingException e) {
            return Optional.empty();
        }
    }

    public <T> Optional<T> convertValue(Object fromValue, Class<T> toValueType) {
        return Optional.of(MAPPER.convertValue(fromValue, toValueType));
    }

    public <T> Optional<T> convertValue(Object fromValue, TypeReference<T> toValueTypeRef) {
        return Optional.of(MAPPER.convertValue(fromValue, toValueTypeRef));
    }

}

File: src/main/java/com/example/oauth/util/JsonUtils.kt
package com.example.oauth.util

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper

class JsonUtils private constructor() {

    init {
        throw UnsupportedOperationException("This is a utility class and cannot be instantiated")
    }

    companion object {
        private val MAPPER = ObjectMapper()

        fun <T> readValue(content: String, valueType: Class<T>): T? {
            return try {
                MAPPER.readValue(content, valueType)
            } catch (e: JsonProcessingException) {
                null
            }
        }

        fun <T> readValue(content: String, valueTypeRef: TypeReference<T>): T? {
            return try {
                MAPPER.readValue(content, valueTypeRef)
            } catch (e: JsonProcessingException) {
                null
            }
        }

        fun writeValue(value: Any): String? {
            return try {
                MAPPER.writeValueAsString(value)
            } catch (e: JsonProcessingException) {
                null
            }
        }

        fun <T> convertValue(fromValue: Any, toValueType: Class<T>): T {
            return MAPPER.convertValue(fromValue, toValueType)
        }

        fun <T> convertValue(fromValue: Any, toValueTypeRef: TypeReference<T>): T {
            return MAPPER.convertValue(fromValue, toValueTypeRef)
        }
    }

}